firedrake/meshadapt.py

```diff
@@ -0,0 +1,47 @@
+from __future__ import absolute_import, print_function, division
+
+import numpy as np
+
+from firedrake.mesh import Mesh
+import firedrake.dmplex as dmplex
+from firedrake.petsc import PETSc
+
+
+__all__ = ['adapt']
+
+
+
+
+def adapt(mesh, metric):
+    """ Adapt the mesh to a prescribed metric field.
+
+    :arg mesh: the base mesh to adapt
+    :arg metric: a metric tensor field (a Function of a TensorFunctionSpace)
+
+    :return: a new mesh adapted to the metric
+    """
+    
+    dim = mesh._topological_dimension
+    entity_dofs = np.zeros(dim+1, dtype=np.int32)
+    entity_dofs[0] = mesh.geometric_dimension()
+    coordSection = mesh._plex.createSection([1], entity_dofs, perm=mesh.topology._plex_renumbering)
+    
+    plex = mesh._plex
+    vStart, vEnd = plex.getDepthStratum(0)
+    nbrVer = vEnd - vStart
+    
+    dmCoords = mesh.topology._plex.getCoordinateDM()
+    dmCoords.setDefaultSection(coordSection)    
+
+    with mesh.coordinates.dat.vec_ro as coords:
+        mesh.topology._plex.setCoordinatesLocal(coords)
+    with metric.dat.vec_ro as vec:
+    	dmplex.reorder_metric(plex, vec, coordSection)
```

Comment:
You asked for a read-only view of the metric, but then update it in place.  Don't do that!  Instead, return a new reordered Vec in `reorder_metric`.

Plausibly for simplification, you could also use the generic reordering to reorder coordinates to send into the DM as well?  Then this would look like

```python
def adapt(mesh, metric):
    assert mesh.coordinates in P1
    assert metric in P1
    with metric.dat.vec_ro as v:
        reordered_metric = to_petsc_numbering(v, metric.function_space())
    with mesh.coordinates.dat.vec_ro as v:
        reordered_coords = to_petsc_numbering(v, mesh.coordinates.function_space())
    dm = mesh._plex
    dm.setCoordinates(reordered_coords)
    newdm = dm.adapt(reordered_metric)
    # distribute=True needs to be set once parallel redistribution works
    return Mesh(newdm, distribute=False)
```